esphome/components/uart/__init__.py

```diff
@@ -95,6 +98,20 @@ def validate_invert_esp32(config):
     return config
 
 
+def validate_port_name(config):
+    if CORE.is_host:
+        if CONF_TX_PIN in config or CONF_RX_PIN in config:
+            raise cv.Invalid(
+                "TX and RX pins are not supported for UART on host platform."
+            )
+        if CONF_NAME not in config:
+            raise cv.Invalid("Name is required for UART on host platform.")
+    else:
+        if CONF_NAME in config:
+            raise cv.Invalid("Name is not supported for UART on non-host platforms.")
+    return config
```

Comment:
With the changes to the field itself, this can be thinned out. 

The `else` is no longer required because of `only_on()`, and the `if CONF_NAME not in config:` will never be reached because of `cv.has_at_least_one_key`

Technically we could put `cv.only_on` on the TX/RX fields to, but that could get messy.

```suggestion
def validate_port_name(config):
    if CORE.is_host:
        if CONF_TX_PIN in config or CONF_RX_PIN in config:
            raise cv.Invalid(
                "TX and RX pins are not supported for UART on host platform."
            )
    return config
```